bundles/org.openhab.binding.magentatv/src/main/java/org/openhab/binding/magentatv/internal/handler/MagentaTVHandler.java

```diff
@@ -0,0 +1,682 @@
+            @Override
+            public void run() {
+                renewEventSubscription();
+            }
+        }, 2, 5, TimeUnit.MINUTES);
+    }
+
+    /**
+     * Thing initialization:
+     * - initialize thing status from UPnP discovery, thing config, local network settings
+     * - perform Oath if userID is not configured and credentials are available
+     * - wait for NotifyServlet to initialize (solves timing issues on fast startup)
+     */
+    @SuppressWarnings("null")
+    @Override
+    public void initialize() {
+        // The framework requires you to return from this method quickly. For that the initialization itself is executed
+        // asynchronously
+        logger.debug("Initialize Thing...");
+        // updateStatus(ThingStatus.UNKNOWN);
+        scheduler.execute(() -> {
+            String errorMessage = "";
+            try {
+                // Example for background initialization:
+                // All relevant parameters will be derived from the thing config
+                // the final result will be saved to the thing properties and can be viewed in
+                // PaperUI
+                thingConfig.initializeConfig(getConfig().getProperties());
+                Map<String, Object> discoveredProperties = handlerFactory.getDiscoveredProperties(thingConfig.getUDN());
+                if (discoveredProperties != null) {
+                    thingConfig.updateConfig(discoveredProperties); // get network parameters from control
+                }
+                if (thingConfig.getUDN().isEmpty()) {
+                    // get UDN from device name
+                    String uid = this.getThing().getUID().getAsString();
+                    thingConfig.setUDN(StringUtils.substringAfterLast(uid, ":"));
+                }
+                if (thingConfig.getMacAddress().isEmpty()) {
+                    // get MAC address from UDN (last 12 digits)
+                    String macAddress = StringUtils.substringAfterLast(thingConfig.getUDN(), "_");
+                    if (macAddress.isEmpty()) {
+                        macAddress = StringUtils.substringAfterLast(thingConfig.getUDN(), "-");
+                    }
+                    thingConfig.setMacAddress(macAddress);
+                }
+
+                control = new MagentaTVControl(thingConfig, network);
+                thingConfig.updateConfig(control.getConfig().getProperties()); // get network parameters from control
+                authenticateUser();
+
+                // wait for NotifyServlet to initialze
+                if (!handlerFactory.getNotifyServletStatus()) {
+                    logger.debug("Waiting on NotifyServlet to start...");
+                    int iRetries = 30;
+                    while ((iRetries-- > 0) && !handlerFactory.getNotifyServletStatus()) {
+                        logger.trace("Waiting for init, {} sec remaining", iRetries);
+                        Thread.sleep(1000);
+                    }
```

Comment:
Don't use Thread.sleep wait in a loop to check for a status. Use either a wait/notify, lock/signal, or CompletableFuture instead.

Reply:
removed